File: dm_ws/src/dm_motor_controller/scripts/publish_trajectory.py
#!/usr/bin/env python3
import rospy
import yaml
from dm_motor_controller.msg import MITCommand

def main():
    rospy.init_node('trajectory_publisher')

    pub = rospy.Publisher('/dm_motor/mit_command', MITCommand, queue_size=10)
    rospy.sleep(1.0)

    path = rospy.get_param('~trajectory_file', 'config/trajectory.yaml')
    target_mode = rospy.get_param('~mode', 'p')  # 'p' for position, 'v' for velocity

    with open(path, 'r') as f:
        traj_data = yaml.safe_load(f)

    filtered = [p for p in traj_data['params'] if p['mode'] == target_mode]
    if not filtered:
        rospy.logwarn("No points found for mode '%s'", target_mode)
        return

    rate = rospy.Rate(200)  # default 200hz
    for point in filtered:
        cmd = MITCommand()
        cmd.kp = point['kp']
        cmd.kd = point['kd']
        cmd.position = point['position']
        cmd.velocity = point['velocity']
        cmd.torque = point['torque']
        rospy.loginfo("Mode [%s] → Sending: %s", target_mode, cmd)
        pub.publish(cmd)
        rate.sleep()

if __name__ == '__main__':
    main()

File: dm_ws/src/dm_motor_controller/src/publish_trajectory.cpp
#include <ros/ros.h>
#include <dm_motor_controller/MITCommand.h>
#include <yaml-cpp/yaml.h>
#include <fstream>

int main(int argc, char** argv) {
    ros::init(argc, argv, "trajectory_publisher");
    ros::NodeHandle nh("~");

    ros::Publisher pub = nh.advertise<dm_motor_controller::MITCommand>("/dm_motor/mit_command", 10);
    ros::Duration(1.0).sleep();

    std::string file_path, mode;
    nh.param<std::string>("trajectory_file", file_path, "config/trajectory.yaml");
    nh.param<std::string>("mode", mode, "p");

    YAML::Node config = YAML::LoadFile(file_path);


    double frequency = config["rate"] ? config["rate"].as<double>() : 1.0;
    if (frequency <= 0.0) {
        ROS_WARN("Invalid rate in YAML, defaulting to 1.0 Hz");
        frequency = 1.0;
    }

    auto params = config["params"];
    std::vector<dm_motor_controller::MITCommand> traj;
    for (const auto& point : params) {
        if (point["mode"].as<std::string>() != mode) continue;

        dm_motor_controller::MITCommand cmd;
        cmd.kp = point["kp"].as<float>();
        cmd.kd = point["kd"].as<float>();
        cmd.position = point["position"].as<float>();
        cmd.velocity = point["velocity"].as<float>();
        cmd.torque = point["torque"].as<float>();
        traj.push_back(cmd);
    }

    if (traj.empty()) {
        ROS_WARN("No valid trajectory points found for mode '%s'", mode.c_str());
        return 0;
    }

    std::string mode_str = (mode == "p") ? "Position Control" :
                           (mode == "v") ? "Velocity Control" : "Unknown";
    ROS_INFO("Running in [%s] mode with %lu trajectory points at %.2f Hz.", mode_str.c_str(), traj.size(), frequency);

    ros::Rate rate(frequency);
    for (size_t i = 0; i < traj.size() && ros::ok(); ++i) {
        ROS_INFO("Sending point %lu/%lu: kp=%.2f, kd=%.2f, pos=%.2f, vel=%.2f, tq=%.2f",
                 i+1, traj.size(), traj[i].kp, traj[i].kd, traj[i].position,
                 traj[i].velocity, traj[i].torque);
        pub.publish(traj[i]);
        rate.sleep();
    }

    return 0;
}
